test/Spec/BotPlutusInterface/Balance.hs

```diff
@@ -105,3 +165,131 @@ addUtxosForChange = do
   case ebalancedTx of
     Left e -> assertFailure (Text.unpack e)
     Right balancedTx -> txInputs <$> balancedTx @?= Right (Set.fromList [txIn1, txIn2])
+
+dontAddChangeToDatum :: Assertion
+dontAddChangeToDatum = do
+  let scrTxOut' =
+        ScriptChainIndexTxOut
+          valAddr
+          (Right validator) -- (valHash, Just validator)
+          (Right scrDatum) -- (scrDatumHash, Just scrDatum)
+          scrValue
+      scrTxOut = Ledger.toTxOut scrTxOut'
+      usrTxOut' =
+        PublicKeyChainIndexTxOut
+          pkhAddr3
+          (Ada.lovelaceValueOf 1_001_000)
+      usrTxOut = Ledger.toTxOut usrTxOut'
+      -- initState :: MockContractState ()
+      initState =
+        def & utxos .~ [(txOutRef6, scrTxOut), (txOutRef7, usrTxOut)]
+          & contractEnv .~ contractEnv'
+      pabConf :: PABConfig
+      pabConf = def {pcOwnPubKeyHash = pkh3}
+      -- contractEnv' :: ContractEnvironment ()
+      contractEnv' = def {cePABConfig = pabConf}
+
+      -- Input UTxOs:
+      -- UTxO 1:
+      -- - From: User
+      -- - Amt : 1.001 ADA
+      -- UTxO 2:
+      -- - From: Script
+      -- - Amt : 0.5 ADA + 200 Tokens
+      --
+      -- Output UTxOs:
+      -- UTxO 1:
+      -- - To : User
+      -- - Amt: 1 ADA
+      -- UTxO 2:
+      -- - To : Script
+      -- - Amt: 1.0005 Ada + 200 Token
+      --
+      -- Fees   : 400 Lovelace
+      -- Change : 100 Lovelace
+
+      scrLkups =
+        Constraints.unspentOutputs (Map.fromList [(txOutRef6, scrTxOut'), (txOutRef7, usrTxOut')])
+          <> Constraints.ownPaymentPubKeyHash paymentPkh3
+      txConsts =
+        -- Pay the same datum to the script, but with more ada.
+        Constraints.mustPayToOtherScript valHash scrDatum (scrValue <> Ada.lovelaceValueOf 500)
+          -- <> Constraints.mustPayToOtherScript valHash scrDatum (Ada.lovelaceValueOf 1_000_000)
+          <> Constraints.mustPayToPubKey paymentPkh3 (Ada.lovelaceValueOf 1_000_000)
+          <> Constraints.mustSpendScriptOutput txOutRef6 Ledger.unitRedeemer
+          <> Constraints.mustSpendPubKeyOutput txOutRef7
+      eunbalancedTx = Constraints.mkTx @Void scrLkups txConsts
+
+  case eunbalancedTx of
+    Left mkTxErr -> assertFailure ("MkTx Error: " <> show mkTxErr)
+    Right unbalancedTx -> do
+      let (eRslt, _finalState) = runPABEffectPure initState (balanceTxIO @() @'[PABEffect ()] pabConf pkh3 unbalancedTx)
+      case eRslt of
+        (Left txt) -> assertFailure ("PAB effect error: " <> Text.unpack txt)
+        (Right (Left txt)) -> assertFailure $ "Balancing error: " <> Text.unpack txt -- <> "\n(Tx: " <> show unbalancedTx <> ")"
+        (Right (Right trx)) -> do
+          let scrTxOut'' = scrTxOut' & Ledger.ciTxOutValue <>~ Ada.lovelaceValueOf 500
+              scrTxOutNew = Ledger.toTxOut scrTxOut''
+          assertBool
+            ( "Expected UTxO not in output Tx."
```

Comment:
I think for readability and debuggability it would be better to find script UTxO by `TxOutRef` and compare with expected, so in fail message we can print something like
"Script output value changed unexpectedly. Expecting [value of expected output], but got [real value]"

Reply:
I dunno, I like being able to see where all the UTxOs are going and what all the outputs are; it makes it easier to see *how* the transaction is being balanced.

Reply:
I feel like we can use some binds here, and case over the final result, since all your failure cases lead to `assertFailure`.
So, instead of an explicit case over, say, `eunbalancedTx`, you bind and and `first` your `e -> String` function onto it

Reply:
Also, consider using `printf` in your assertion errors

Reply:
Okay, I just wrote a simple `liftAssertFailure` function that takes an `Either` and runs `assertFailure` if it's a `Left` and just returns the value if it's a `Right` (since `assertFailure` throws an error). 

As for using `printf`, I feel like that would make some of my lines too long; I like the way the "Expected UTxOs" lines line up. I might use it for the "Other UTxOs don't contain the right change" lines, though.